src/HooglePlus/FilterTest.hs

```diff
@@ -168,4 +171,30 @@ handleAnyException = (`catch` (return . handler . show :: SomeException -> IO (M
   where
     handler ex  | "timeout" `isInfixOf` ex = Nothing
                 | otherwise = (Just . Left. UnknownError) ex
-fmtFunction_ = printf "((%s) :: %s) %s" :: String -> String -> String -> String
\ No newline at end of file
+fmtFunction_ = printf "((%s) :: %s) %s" :: String -> String -> String -> String
+
+checkDuplicates :: MonadIO m => [String] -> String -> String -> FilterTest m Bool
+checkDuplicates modules sigStr body = do
+  st <- get
+  case sampleResults st of
+    Nothing -> do
+      result' <- liftIO (Just . (`SampleResult` []) <$> generateInputs modules funcSig)
+      modify $ \st -> st { sampleResults = result' }
+      checkDuplicates modules sigStr body
+    Just (SampleResult inputs results) -> do
+      evals <- liftIO $ evalResults inputs funcSig body
+      let isDuplicated = evals `elem` results
+      let results' = if isDuplicated then results else evals:results
+
+      modify $ \st -> st { sampleResults = Just $ SampleResult inputs results' }
+      return $ not isDuplicated
+
+  where
+    funcSig = (instantiateSignature . parseTypeString) sigStr
+
+    generateInputs :: [String] -> FunctionSignature -> IO [String]
```

Comment:
This seems like a reasonably important function that we'll want to be able to test in isolation. I would raise it to a top-level declaration. This is the function that produces random sets of inputs, right?

Reply:
Exactly. But it doesn't really do more things than calling the top-level _random input generation_ function multiple times to form a list.